### DISCUSSION TITLE
M5 Stick C Plus 2 support added

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#641] [Announcements]
I bought a few M5 Stick C Plus 2 units and found they weren't directly compatible with the other M5 Stick C Plus projects.  So I did two things:

1) First, I added support for a spectrum2 project that uses the Plus2 module, fixed it up, and got it working
2) Next, I refactored all M5 support for the Stick, Stick Plus, Stick Plus 2, Core Stack 2, and so on:
     - I switched to the M5 Unified library and revised our code to use it, so it's almost the same in all cases

Defining any of the things like M5STICKCPLUS will wind up defining USE_M5, and that's now what most decisions are based on rather than individual tests.

The audio recording format is also different, but it works across all chips, so for non-mesmerizer cases, I switched to M5.Mic.Record() instead of directly doing i2s audio.  

I've checked it into my own fork, if you want it immediately you can grab it there, otherwise I'll let it percolate a bit and then migrate to main in a little bit.

 =Dave

**Comment by davepl** [comment]
BTW, on the plus side (or Plus2 side), the screen is much faster than the original Plus.  It runs the little spectrum display at 142 fps!

It also has 2MB of PSRAM, which I've enabled for all Plus2 projects.

**Comment by robertlipe** [comment]
What's the plan for audio on non-espressif boards?

**Comment by davepl** [comment]
Non-esp32?  Don’t care.  It works on everything it did a week ago, plus the Plus2 now.  What board are you trying to support?

- Dave

**Comment by robertlipe** [comment]
Of course still the subset of esp32 this project always supported.
Espressif owns m5stack. What about the boards using the esp32s that dont
likely use anything called the "M5 unified library"?

https://www.espressif.com/en/news/Espressif_Acquires_M5Stack

**Comment by davepl** [comment]
Not to be pedantic, but my question:  

>  It works on everything it did a week ago, plus the 
>  Plus2 now. What board are you trying to support? 

It requires an espressif chip, not their module.  But it supports their modules.

- Dave

**Comment by robertlipe** [comment]
My question could have been phrased better for those not believing a name
change is imminent. More precisely, my original question was does your
statement

 for non-mesmerizer cases, I switched to M5.Mic.Record() instead of
directly doing i2s audio.

Apply to all devices or does i2s remain for others?

So, AINUI, m5s were changed to use the m5 path and everything else uses
what it's always used. Mesmerizer was presumably never on that m5 path, so
I don't know why it gets the shout-out.

I probably should have just waited to get home and reviewed the change
instead of your description, but that wouldn't have helped the rest of the
audience.

**Comment by davepl** [comment]
I have fixed the M5 audio issues, and everything works nicely now.  Tested on M5 Stick C, M5 Stick C Plus 2, M5Stack, and Mesermizer.

(Very) long story short, the raw audio data that comes back from the M5 mic is signed data centered around zero, whereas the mic in the mesmerize returns unsigned 16 bit values.

I also simplified the NOISE_FLOOR and NOISE_CUTOFF stuff, so it's only defined once, and added an AUDIO_MIC_SCALAR that takes car of adjusting the relative sensitivities of the mics.

It's in a PR that should go in soon.

**Comment by Witchdoc59** [comment]
I ended up getting an M5 Stick C Plus 2 since the M5 Stick C Plus was no longer available.   I downloaded the latest build of the NightDriverStrip software.  Managed to get it flashed to the M5 but it's not working very well.   I have a 5 meter strip.  Not sure of how many LEDs.   Down the length of the strip I seem to have random red leds lighting up, not flashing or anything just a constant red light.   I do get some flashing at the start of the strip.  Clearing my throat will light up or saying BOO will light up the first 8 LEDs.  It is a 5 volt strip and I have a 5 volt power supply going to it.   So my question is, is there some place in the code where I have to define the number of LEDs I'm lighting up?

**Comment by robertlipe** [comment]
A 5m 5v strip will need multiple power injections to light it at
non-trivial brightness.

Yes, you must set the number of pixels for your effect in inc*/globals.h

Remember that for some strips, pixels and bulbs aren't the same.

**Reply by Witchdoc59** [reply]
Thanks for the fast response.   The power block I have is 5 volts at 6 amps.   Do I need more than that for the 5 meter strip?  There are 60 RGB LEDs per meter.  Would that be 300 per strip or is each color counted making it 180 per meter?  Can you tell me where in globals.h I find the number of pixels. I'm trying to get the Laserline to work as in Dave's video.

**Reply by rbergen** [reply]
I'll leave the power question to @robertlipe, as I have less experience with strips of different sizes (technically, I have no experience with actual strips at all, only matrices).

However, the code change question I can answer.

To make the Laserline project work on the M5StickC Plus2, changes have to be made in 2 files. You may already have done the first to get to where you are now, I'm just adding it here for completeness:

- In platformio.ini, change the laserline project ("env") to use the M5StickC Plus2 instead of the "regular" Plus. You do this by changing the `[env:laserline]` block ([lines 598 to 602 in the current `main` branch](https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/platformio.ini#L598-L602) so that it looks like this:
  
  ```ini
  [env:laserline]
  extends         = dev_m5stick-c-plus2
  build_flags     = -DLASERLINE=1
                    ${remote_flags.build_flags}
                    ${dev_m5stick-c-plus2.build_flags}
  ```

  Note that the change has to be made in _two_ lines: the `extends` line, and the last line of the `build_flags` sequence (also the last line in the block).

- In globals.h, change the definition for `MATRIX_WIDTH` in the `LASERLINE` block to the total number of RGB LEDs in your strip. Which would indeed be 300 if you have 5 meters of 60 RGB LEDs per meter. In the current `main` branch [the `LASERLINE` block starts at line 499](https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/6cf79ee256c3218d7493f597c8f4c4a72ad896e1/include/globals.h#L499), and [the line to change is line 522](https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/6cf79ee256c3218d7493f597c8f4c4a72ad896e1/include/globals.h#L522) (change the current value of 700 to 300).

**Reply by Witchdoc59** [reply]
Thanks again for the reply.  Dave must have made the change to platformio.ini since it is already set to use the dev_m5stick-c-plus2

**Reply by rbergen** [reply]
No, he didn't. Not for the Laserline project. 

The code as it comes "off the shelf" currently has two platformio.ini environments set up to use the M5StickC Plus2: `spectrum2` (which flashes the `SPECTRUM` project), and `pdpgrid` (which selects the `PDPGRID` project). 

To use the Laserline project on the M5StickC Plus2, you really have to make the changes to platformio.ini yourself as I stated them, and tell PlatformIO to flash the `laserline` environment. 
Otherwise chances are high that you end up with some iteration of another project flashed on your device, in a way that is not actually compatible with the M5StickC Plus2.

**Reply by Witchdoc59** [reply]
Thanks, upon closer inspection I see I did have to make a change.  I was missing the 2.

**Comment by rbergen** [comment]
@Witchdoc59 Also, what project are you flashing on it?

**Reply by Witchdoc59** [reply]
I'm trying to duplicate what Dave did in the video where he created the laser cannon.  I bought all the stuff he used in the video.  I did buy the wrong M5 initially, I didn't have the 'plus'.   The fact that I am getting some response, makes me think I've got the wiring correct.

**Reply by rbergen** [reply]
Wiring is one thing (and you may well have got that right if you see some LED activity), setting up the configuration in platformio.ini and globals.h for the right device and your particular LED setup is another. Which is what I explained in the comment I left earlier today.

As I know all too well by personal experience, there is a bit of a learning curve to all this. Once you get your head around the platformio.ini/globals.h combo for the device and project you're trying to flash and get working, things will become a lot easier. We're happy to help you get there.

**Comment by Witchdoc59** [comment]
So all the random red leds have disappeared.  With certain loud noices or frequencies the first 8 LEDs will flash and they flash different colors.  The device only works if I have it plugged into the USB-C which makes me think there may be a problem with the wiring.

**Reply by rbergen** [reply]
As I already said, I'm not the expert when it comes to properly hooking up strips to power. But the M5StickC Plus2 does have a battery and should run for a decent while on it, provided it only has to drive the LEDs' data signal. For that to be the case, the LEDs in the strip indeed have to be powered through a dedicated power supply.

If the whole thing only works when the M5StickC Plus2 is connected to USB-C, then I think you may actually be trying to drive your LEDs from the M5StickC Plus2, or effectively from whatever the USB-C power source is. If that's a laptop or even desktop computer, I'd recommend disconnecting it **immediately** and sorting out the strip's power feed before you proceed, because you could well be working towards your computer's USB-C port blowing up - and generating some magic smoke in the process.

Maybe you can send some pictures of your setup, and particularly the wiring between the M5StickC Plus2 and the strip on the one hand, and the strip and its power supply on the other?

**Reply by Witchdoc59** [reply]
![20241105_055648 - Copy](https://github.com/user-attachments/assets/bd2d4b44-7867-4c7a-b583-9e222ba6d1ce)
![20241105_055648](https://github.com/user-attachments/assets/a27d5f87-29e6-4ea1-8fce-2e2c1936335d)
![20241105_055521 - Copy](https://github.com/user-attachments/assets/782011e7-b16d-4bb5-8982-78f7f6625da9)
![20241105_055521](https://github.com/user-attachments/assets/cf1406f5-d650-48f3-a575-cff3be9eaeba)

**Reply by Witchdoc59** [reply]
https://github.com/user-attachments/assets/62d9ca62-6e77-4a0c-975b-a3a23898b7ca

**Comment by davepl** [comment]
Just FWIW, the battery on the M5 can run a few LEDs, but it fails with even 20 of them at reasonable brightness (actually cooked the regulator in mine!)

- Dave

**Reply by rbergen** [reply]
Thanks for confirming that. As I understand it, @Witchdoc59 is trying to drive a 5 meter strip with a total of 300 LEDs, and that's quite clearly beyond what the M5Stick C Plus2's battery can handle.

**Reply by Witchdoc59** [reply]
No, I'm not trying to drive the strip with the M5.   I bought all the items that Dave listed in his video.   Including a single power brick that is 5V 6A.   And at the moment I've only got 1 of the 2 strips hooked up.

**Reply by rbergen** [reply]
To clarify: I didn't mean to imply that you were definitively driving 300 LEDs from the M5StickC Plus2 (let alone _trying_ to do so), just that you are trying to drive 300 LEDs and the M5 can't push that.

What are the doubts about the wiring you mentioned yourself, before?

**Reply by Witchdoc59** [reply]
If I disconnect the USB-C cable from the M5 then the battery eventually dies.  I am thinking that the system should be able to run without the USB-C connection.

**Reply by rbergen** [reply]
It should indeed, for a while. But "eventually" is a rather flexible term. :) 

Would it be possible for you to post some pictures of your wiring setup?

**Reply by Witchdoc59** [reply]
When I get home from work I'll take some photos

**Comment by davepl** [comment]
I’d say so :-). With 5M, you’d want to inject a fresh 5V source from a larger gauge cable ever 2M at LEAST, and ideally ever 1M if you’re going to use white, etc.

- Dave

**Comment by davepl** [comment]
Just FWIW, you can power the M5 Stick C Plus from the USB connector or the pinset on the end, but not through the grove connector.

_ Dave

**Reply by Witchdoc59** [reply]
ahh, I was assuming that the grove connector provided power from the power brick.
